book/accessibility.md

```diff
@@ -1610,166 +1653,287 @@ used, so let's bind turning it on to the `ctrl-a` keystroke:
+        self.accessibility_focus = None
+```
 
-    def render(self):
+We don't just want to speak this node on every frame, so we'll also
+store the last-spoken node:
+
+``` {.python}
+class Tab:
+    def __init__(self, browser):
         # ...
-        if self.needs_layout:
-            self.document = DocumentLayout(self.nodes)
-            self.document.layout(self.zoom)
-            if self.accessibility_is_on:
-                self.needs_accessibility = True
-            else:
-                self.needs_paint = True
-            self.needs_layout = False
+        self.accessibility_spoken = None
+```
+
+Now we can speak a node exactly when the current accessibility focus
+isn't the last spoken node:
 
+``` {.python}
+class Tab:
+    def speak_update(self):
+        if self.accessibility_focus and \
+            self.accessibility_focus != self.accessibility_spoken:
+            
+            speak_text(self.accessibility_focus.description())
+            self.accessibility_spoken = self.accessibility_focus
+```
+
+Since `accessibility_focus` and `accessibility_spoken` are pointers
+into the accessibility tree, we need to update them when we rebuild
+the accessibility tree. We can do this by finding the new
+accessibility node pointing to the same HTML element as the old one:
+
+``` {.python}
+class Tab:
+    def find_anode(self, node):
+        if not node: return None
+        anodes = [
+            anode for anode in tree_to_list(self.accessibility_tree, [])
+            if anode.node == node
+        ]
+        return anodes[0] if anodes else None
+
+    def render(self):
         if self.needs_accessibility:
+            old_afocus = self.accessibility_focus
+            old_aspoken = self.accessibility_spoken
             self.accessibility_tree = AccessibilityNode(self.nodes)
             self.accessibility_tree.build()
-            self.needs_accessibility = False
-            self.needs_paint = True
-            self.speak_task()
+            if old_afocus:
+                self.accessibility_focus = self.find_anode(old_afocus.node)
+            if old_aspoken:
+                self.accessibility_spoken = self.find_anode(old_aspoken.node)
+            # ...
 ```
 
-The `build` method on `AccessibilityNode` recursively creates the tree. Whether
-a node is represented in the accessibility tree depends on its tag and style,
-which in turn determine not only whether the element is presentational, but if
-not what its semantics are. For example, a `<div>` by itself is by default
-considered presentational or otherwise having no accessibility semantics, but
-`<input>`, `<a>` and `<button>` have them by default. The semantics are called
-the *role* of the element---the role it plays in the meaning of the document.
-And these roles are not arbitrary text; they are specified in a
-[standard][aria-roles] just like rendering.[^not-exposed]
+Now we can add navigation key bindings for the accessibility tree.
```

Comment:
Does really a11y actually work like this? I thought it navigated conceptually by landmarks.

Reply:
Well... There are lots of navigation options in most screen readers, and landmarks are often more efficient than directly navigating the tree. But this one is supported at least by VoiceOver (which is the only screen reader I know), though it's probably not the most common feature.

I guess I'm dissuading myself that this was a good example. Why implement an obscure option instead of the most common one? Because we want to demonstrate the accessibility tree, sure, but surely there is a better reason to build the accessibility tree than supporting this one obscure option? I think this brings us back to the goal stated in the PR: we need better motivation for the accessibility tree being a tree.

Reply:
I would rather we implement important features and not obscure ones.

I added the hover feature a couple of months ago precisely for this reason -- it's an important a11y feature, and "makes use of" the a11y tree in a real way, in the sense that it hit tests it.

The a11y tree could probably have been built with notations on the layout tree about which nodes to skip, etc. But it does seem a bit cleaner to separate the concepts, and also note that it does become more hairy and complicated the more features you add. It's kind of like how in the early days, there probably was not a layout tree and instead browsers hung metadata off of DOM nodes. But nowadays that feels crazy because layout has become complex.